Getting rid of Errno and WinError

waj opened the pull request:
This PR removes the usage of `Errno` and `WinError` as exceptions and instead provides some new classes to distinguish different kind of errors.

This is the hierarchy of the new or updated exceptions:
```
Exception
+- RuntimeError
+- IO::Error
   +- IO::TimeoutError
   +- File::Error
   |  +- File::NotFoundError
   |  +- File::AccessDeniedError
   |  +- File::AlreadyExistsError
   +- Socket::Error
      +- Socket::ConnectError
      +- Socket::BindError
```
More subclasses might be added in the future but I think this is a good starting point.


  * `Exception` now has an optional `os_error` containing the original error code that anyone can use in case the exception class hierarchy is not enough for proper error handling
  * `IO::Timeout` is renamed to `IO::TimeoutError`
  * `Errno` and `Winerror` are now enums
  * `Exception.from_errno` and `Exception.from_winerror` eases creation of exceptions from OS errors
  * `RuntimeError` is used, often when there is a system call that shouldn't fail in normal conditions

(closes #8827)

straight-shoota commented:
I don't like having `os_error` on `Exception`, spilling into the generic exception namespace when it's use is actually limited to a very small amount of exception types from stdlib and maybe a few exceptions in shards with C bindings. This propert doesn't make sense for the vast majority of exceptions in a typical Crystal program.
Can we only add it selectively on the exception types that use it?

Fryguy commented on src/exception.cr:
```diff
@@ -67,6 +73,56 @@ class Exception
     io.flush
   end
+
+  protected def os_error=(@os_error)
```
`:nodoc:` ?  (Not sure we nodoc protected methods, but I saw some elsewhere that had it)

waj replied:
Done, and moved to `SystemError` module

waj commented:
@straight-shoota actually I like it because it ensures uniformity. Many different new exceptions from shards or std might be created from system errors, and so many errors come from the system. I think we should properly document the `from_errno` and `from_winerror` methods and encourage their use. And to be honest, I don't feel that adding a field with a value type is spilling the API. It's convenient and unnoticed most of the time.

jhass commented:
As a compromise we could also have `module SystemError` (or `SyscallError` or so) for it.

waj commented:
Ok, agreed. I moved the implementation to the `SystemError` module and included in `RuntimeError` and `IO::Error`. Also added some docs to that module 🙂

bcardiff commented on src/crystal/system/win32/file.cr:
```diff
@@ -44,12 +44,12 @@ module Crystal::System::File
       return nil
     else
-      raise WinError.new(func_name, error)
+      raise ::File::Error.from_winerror(message, file: path)
```
The `error` variable should be used in this call

bcardiff commented on src/socket.cr:
````diff
@@ -179,15 +195,15 @@ class Socket < IO
   # sock.bind Socket::IPAddress.new("192.168.1.25", 80)
   # ```
-  def bind(addr)
+  private def bind(addr, addrstr)
````
It seems we lost `sock.bind Socket::IPAddress.new("192.168.1.25", 80)` since the method is now private. A miss edition probably/

oprypin commented:
At a glance, this seems strictly better than https://github.com/crystal-lang/crystal/pull/5003 
👍

waj commented:
Thanks @oprypin. I considered some of the ideas and the feedback received on that PR.

RX14 commented on src/file/error.cr:
```diff
@@ -0,0 +1,41 @@
+end
+
+class File::Error < IO::Error
+  getter file : String
```
Would be nice to have some documentation on these.

waj replied:
Of course. Documentation is still missing in many classes and methods.

RX14 commented on src/process.cr:
```diff
@@ -142,7 +142,7 @@ class Process
       errno = Errno.value
       LibC.pthread_sigmask(LibC::SIG_SETMASK, pointerof(oldmask), nil)
-      raise Errno.new("fork", errno)
+      raise RuntimeError.from_errno("fork", errno)
```
I think it would be kind of nice if `Process.new` on a non-existing file raised `File::NotFoundError`, i.e. all file-related errors were handled like this, even in mixed-function syscalls.

RX14 commented on src/errno.cr:
```diff
@@ -16,232 +16,52 @@ end
+                    EHOSTUNREACH ENOTEMPTY EUSERS EDQUOT ESTALE EREMOTE ENOLCK ENOSYS EOVERFLOW
+                    ECANCELED EIDRM ENOMSG EILSEQ EBADMSG EMULTIHOP ENODATA ENOLINK ENOSR ENOSTR
+                    EPROTO ETIME EOPNOTSUPP ENOTRECOVERABLE EOWNERDEAD) %}
+    {% if LibC.has_constant?(value) %}
```
We should either move this enum into `Crystal::` so it is not public or have it be always the same list of errno despite the platform: this enum must have the same *members* regardless of platform.

waj replied:
But not all errno values exist in all platforms. Maybe we can map those to EINVAL. But there should be at least a warning if you try to use any of those values in a platform where the value doesn't exist.

RX14 replied:
We decided a long time ago that the compile-time API must be identical between platforms so that everything compiles on every platform.

If the errno value doesn't exist on every platform, it just never gets used, never gets raised, so it can be any value.

I don't think we need warnings for this.

aravindavk replied:
The existence of an Error number is one part of the problem. There is platform-specific value for some error codes. For example, `grep -R "EDEADLK" src/lib_c/*`

```
src/lib_c/aarch64-linux-gnu/c/errno.cr:        EDEADLK = 35
src/lib_c/aarch64-linux-musl/c/errno.cr:       EDEADLK = 35
src/lib_c/amd64-unknown-openbsd/c/errno.cr:    EDEADLK = 11
src/lib_c/arm-linux-gnueabihf/c/errno.cr:      EDEADLK = 35
src/lib_c/i386-linux-gnu/c/errno.cr:           EDEADLK = 35
src/lib_c/i386-linux-musl/c/errno.cr:          EDEADLK = 35
src/lib_c/i686-linux-gnu/c/errno.cr:           EDEADLK = 35
src/lib_c/i686-linux-musl/c/errno.cr:          EDEADLK = 35
src/lib_c/x86_64-darwin/c/errno.cr:            EDEADLK = 11
src/lib_c/x86_64-dragonfly/c/errno.cr:         EDEADLK = 11
src/lib_c/x86_64-freebsd/c/errno.cr:           EDEADLK = 11
src/lib_c/x86_64-linux-gnu/c/errno.cr:         EDEADLK = 35
src/lib_c/x86_64-linux-musl/c/errno.cr:        EDEADLK = 35
src/lib_c/x86_64-macosx-darwin/c/errno.cr:     EDEADLK = 11
src/lib_c/x86_64-openbsd/c/errno.cr:           EDEADLK = 11
src/lib_c/x86_64-portbld-freebsd/c/errno.cr:   EDEADLK = 11
src/lib_c/x86_64-unknown-freebsd/c/errno.cr:   EDEADLK = 11
src/lib_c/x86_64-windows-msvc/c/errno.cr:      EDEADLK = 36
```

straight-shoota commented on src/process.cr:
```diff
@@ -479,17 +474,7 @@ class Process
-      io << ")"
-    end
-    raise Errno.new(error_message)
+    raise RuntimeError.from_errno
```
Why is this not `from_error("execvp")`?

waj replied:
I think eventually we should make an effort to improve many error messages. I'm not sure it's a good idea to leak the syscall name that failed. Maybe the same error message we have above: "Error executing process"

straight-shoota replied:
Yes, that could work. "execvp" has always just been a stand-in for a proper message.

Using an explicit message here would make `from_errno` without arguments obsolete. I think it should be required to provide at least some message for context information.

didactic-drunk replied:
Is it leaking the syscall name or telling you exactly what failed, which manpage to read and what to search for?

straight-shoota replied:
I actually don't think leaking the name is an issue. It's really pretty concise about defining the context of the error. (Btw. it's a libc call, not a syscall)

But in other cases, I think we can perhaps be more expressive about the error context. There's a huge problem space with localization, though. So IDK, but using just the call looks like a pretty reasonable solution, at least for the time being.
